torch/jit/__init__.py

```diff
@@ -702,32 +682,100 @@ def init_then_register(self, *args, **kwargs):
         return super(ScriptMeta, cls).__init__(name, bases, attrs)
 
 
-class ScriptModule(with_metaclass(ScriptMeta, torch._C.ScriptModule)):
-
-    def __setattr__(self, name, value):
-        if isinstance(value, Parameter):
-            self._register_or_set_parameter(name, value)
-        elif isinstance(value, ScriptModule):
-            self._register_module(name, value)
-            # note: script modules are subclassed in python and the
-            # C++ script::Module class will not hold references to them
-            # to ensure that you always get the same python value here
-            # we store it as a native attribute _in addition to_
-            # registering it with the C++ script::Module
-            object.__setattr__(self, name, value)
-        else:
-            object.__setattr__(self, name, value)
+class ScriptModule(with_metaclass(ScriptMeta, Module, torch._C.ScriptModule)):
+    def __init__(self, optimize=True):
+        Module.__init__(self)
+        torch._C.ScriptModule.__init__(self, optimize)
+        self._parameters = OrderedParameterDict(self)
+        self._buffers = OrderedBufferDict(self)
+        self._modules = OrderedModuleDict(self)
 
     def __getattr__(self, attr):
-        r = self._get_attribute(attr)
-        if r is None:
-            raise AttributeError("'{}' object has no attribute '{}'".format(self.__class__.__name__, attr))
-        return r
+        if self._has_method(attr):
+            return self._get_method(attr)
+        return Module.__getattr__(self, attr)
+
+    # Module already has this method defined, so we
+    # need to override it and send it through the ScriptModule lookup
+    def forward(self, *args, **kwargs):
+        return self.__getattr__('forward')(*args, **kwargs)
```

Comment:
Using `__getattr__` is a bit conterintuitive. Can't you just `self._get_method('forward')`? That will be more explicit and faster